python/python/bystro/cli/cli.py

```diff
@@ -0,0 +1,314 @@
+    )
+    signup_parser.add_argument(
+        "--host",
+        required=True,
+        help="Host of the Bystro API server, e.g. https://bystro-dev.emory.edu",
+    )
+    signup_parser.add_argument(
+        "--port",
+        type=int,
+        help=(
+            "Port of the Bystro API server, e.g. 443. "
+            "Defaults to the standard port for the given protocol (http: 80, https: 443)"
+        ),
+    )
+    signup_parser.set_defaults(func=signup_cli)
+
+    user_parser = subparsers.add_parser("get-user", help="Handle user operations")
+    user_parser.set_defaults(func=get_user_cli)
+
+    # Adding the jobs sub-command
+    create_jobs_parser = subparsers.add_parser("create-annotation", help="Create an annotation")
+    create_jobs_parser.add_argument(
+        "--files",
+        nargs="+",
+        required=True,
+        type=str,
+        help="Paths to files: .vcf and .snp formats accepted. Each file is treated as a separate annotation",
+    )
+    create_jobs_parser.add_argument(
+        "--names",
+        nargs="+",
+        type=str,
+        help=(
+            "Names for each job, corresponding to the files provided. "
+            "Must be the same length as the files list. "
+            "Defaults to the basename of the files"
+        ),
+    )
+    create_jobs_parser.add_argument(
+        "--assembly",
+        type=str,
+        required=True,
+        help="Genome assembly (e.g., hg19 or hg38 for human genomes)",
+    )
+    create_jobs_parser.add_argument(
+        "--create-index",
+        type=bool,
+        default=True,
+        help="Whether or not to create a natural language search index the annotation",
+    )
+    create_jobs_parser.set_defaults(func=create_jobs_cli)
+
+    jobs_parser = subparsers.add_parser("get-jobs", help="Fetch one job or a list of jobs")
+    jobs_parser.add_argument("--id", type=str, help="Get a specific job by ID")
+    jobs_parser.add_argument(
+        "--type",
+        choices=list(JOB_TYPE_ROUTE_MAP.keys()),
+        help="Get a list of jobs of a specific type",
+    )
+    jobs_parser.set_defaults(func=get_jobs_cli)
+
+    query_parser = subparsers.add_parser(
+        "query", help="The OpenSearch query string query, e.g. (cadd: >= 20)"
+    )
+    query_parser.add_argument(
+        "--query",
+        required=True,
+        help="The OpenSearch query string query, e.g. (cadd: >= 20)",
+    )
+    query_parser.add_argument("--size", default=10, type=int, help="How many records (default: 10)")
+    query_parser.add_argument(
+        "--from_",
+        default=0,
+        type=int,
+        help="The first record to return from the matching results. Used for pagination",
+    )
+    query_parser.add_argument("--job_id", required=True, type=str, help="The job id to query")
+    query_parser.set_defaults(func=query_cli)
+    proteomics_parser = subparsers.add_parser("proteomics", help="Upload a proteomics dataset")
```

Comment:
I think we talked about putting the proteomics subparser into a separate file, and importing that subparser here. Long term that will be more sustainable that placing everything in this file. See this answer in chatgpt

https://chat.openai.com/share/9ba20cbd-2b6b-4aea-ad60-78e3fa03e861

Reply:
“””
To achieve a CLI (Command Line Interface) setup where you have a main CLI tool (`mycli`) and a subparser for a tool (`tool1`) that lives in a physically different Python file, you can use Python's `argparse` module. The `argparse` module supports the creation of subparsers, which allows you to define commands and subcommands, making it suitable for your use case.

Here's a step-by-step approach to achieve this:

### Step 1: Define the Subparser in a Separate File

First, let's create the subparser for `tool1` in a separate file. Let's call this file `tool1.py`.

```python
# tool1.py
import argparse

def setup_tool1_subparser(subparsers):
    tool1_parser = subparsers.add_parser('tool1', help='Tool1 help')
    tool1_parser.add_argument('tool1cmd', help='Command for tool1')
    tool1_parser.add_argument('--tool1cmdarg1', help='Argument for tool1cmd')
    
    # You can add command function here
    tool1_parser.set_defaults(func=tool1_command)

def tool1_command(args):
    print(f"Executing tool1 with command {args.tool1cmd} and argument {args.tool1cmdarg1}")
```

### Step 2: Main CLI Tool Setup

Next, let's set up the main CLI tool that uses `argparse` to integrate the subparser defined above. We'll call this main file `mycli.py`.

```python
# mycli.py
import argparse
from tool1 import setup_tool1_subparser

def main():
    parser = argparse.ArgumentParser(description='MyCLI tool')
    subparsers = parser.add_subparsers(help='sub-command help')
    
    # Setup subparsers for different tools
    setup_tool1_subparser(subparsers)
    
    # Parse the arguments
    args = parser.parse_args()
    
    if hasattr(args, 'func'):
        args.func(args)
    else:
        # If no subcommands were used, show help
        parser.print_help()

if __name__ == '__main__':
    main()
```

### How It Works

- In `tool1.py`, we define a function `setup_tool1_subparser` that takes `subparsers` as an argument. This function adds a subparser for `tool1` and sets up its arguments.
- In `mycli.py`, we import the `setup_tool1_subparser` function from `tool1.py` and call it, passing the `subparsers` object to it. This way, `mycli` knows about the commands and arguments defined for `tool1`.
- The `tool1_command` function in `tool1.py` is designed to handle the execution logic when `tool1` and its associated commands and arguments are invoked from the CLI.

### Usage

With this setup, you can call `tool1` from your CLI as follows:

```bash
python mycli.py tool1 tool1cmd --tool1cmdarg1 value
```

This command will invoke the `tool1_command` function and print a message indicating that the command and argument have been received and processed.

This approach allows you to maintain a clean separation of concerns between your main CLI tool and its sub-tools, with each sub-tool potentially living in its own file.